ogre2/src/media/materials/programs/thermal_camera_fs.glsl

```diff
@@ -51,28 +52,62 @@ void main()
   float temp = ambient;
   float heatRange = range;
 
-  // get depth
-  float d = getDepth(inPs.uv0);
-
-  // reconstruct 3d viewspace pos from depth
-  vec3 viewSpacePos = inPs.cameraDir * d;
-
-  d = -viewSpacePos.z;
-  d = (d-near) / (far-near);
+  // dNorm value is between [0, 1]. It is used to for varying temperature
+  // value of a fragment.
+  // When dNorm = 1, temp = temp + heatRange*0.5.
+  // When dNorm = 0, temp = temp - heatRange*0.5.
+  float dNorm = 0.5;
 
   // check for heat source
-  float heat = texture(colorTexture, inPs.uv0).x;
+  // heat source are objects with uniform temperature or heat signature
+  // The custom heat source / signature shaders stores heat data in
+  // a vec4 of [heat, 0, 0, 0] so we test to see if it is a heat
+  // source by checking gba == 0. This is more of a hack but the idea is to
+  // avoid having to render an extra pass to create a mask of heat source
+  // objects
+  vec4 rgba = texture(colorTexture, inPs.uv0).rgba;
+  bool isHeatSource = (rgba.g == 0.0 && rgba.b == 0.0 && rgba.a == 0.0);
+  float heat = rgba.r;
+
   if (heat > 0.0)
   {
-    // heat is normalized so convert back to work in kelvin
-    temp = heat * 655.35;
+    if (isHeatSource)
+    {
+      // heat is normalized so convert back to work in kelvin
+      temp = heat * 655.35;
+
+      // set temperature variation for heat source
+      heatRange = heatSourceTempRange;
+    }
+    else
+    {
+      // other non-heat source objects are assigned ambient temperature
+      temp = ambient;
+    }
+  }
 
-    // set temperature variation for heat source
-    heatRange = heatSourceTempRange;
+  // add temperature variation, either as a function of color or depth
+  if (rgbToTemp == 1)
+  {
+    if (heat > 0.0 && !isHeatSource)
+    {
+      // convert to grayscale: darker = warmer
+      float gray = rgba.r * 0.3 + rgba.g * 0.59 + rgba.b*0.11;
```

Comment:
Would it be worth adding a comment about why these particular weights were chosen? I know that there are various RGB -> grayscale conversion formulas out there that use different weights.

Reply:
I rounded the coefficients from here: https://docs.opencv.org/3.4/de/d25/imgproc_color_conversions.html

I noticed there are other ones for gamma sRGB image space here:
https://en.wikipedia.org/wiki/Grayscale